File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.cpp
class Solution {
public:
    int minMoves(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<int>diff(2 * limit + 2, 0);
        for (int i = 0; i < n / 2; i++) {
            int a = min(nums[i], nums[n - 1 - i]);
            int b = max(nums[i], nums[n - 1 - i]);

            diff[2] += 2;
            diff[limit * 2 + 1] -=2;

            diff[a + 1] -=1;
            diff[limit + b + 1] += 1;

            diff[a + b] -= 1;
            diff[a + b + 1] += 1;
        }

        int ans = INT_MAX;
        int cur = 0;
        for (int i = 2; i <= limit * 2; i++) {
            cur += diff[i];
            ans = min(ans, cur);
        }
        return ans;
    }
};

File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.py
class Solution:
    def minMoves(self, nums: List[int], limit: int) -> int:
        n = len(nums)
        diff = [0] * (limit * 2 + 2)

        for i in range(n // 2):
            a = min(nums[i], nums[n - 1 - i])
            b = max(nums[i], nums[n - 1 - i])

            diff[2] += 2
            diff[2 * limit + 1] -= 2

            diff[a + 1] -= 1
            diff[limit + b + 1] += 1

            diff[a + b] -= 1
            diff[a + b + 1] += 1
        
        ans = n
        cur = 0

        for i in range(2, 2 * limit + 1):
            cur += diff[i]
            ans = min(ans, cur)
        
        return ans